aten/src/ATen/ParallelMobileThreadPool.cpp

```diff
@@ -0,0 +1,23 @@
+#include <ATen/ParallelMobileThreadPool.h>
+#include <caffe2/utils/threadpool/ThreadPool.h>
+#include <caffe2/utils/threadpool/pthreadpool.h>
+
+namespace at {
+#ifdef C10_MOBILE
+namespace {
+std::unique_ptr<caffe2::ThreadPool> thread_pool_ = nullptr;
+std::mutex thread_pool_creation_mutex_;
+} // namespace
+pthreadpool_t mobile_threadpool() {
+  std::lock_guard<std::mutex> guard(thread_pool_creation_mutex_);
+  if (!thread_pool_) {
+    thread_pool_ = caffe2::ThreadPool::defaultThreadPool();
+  }
+  return reinterpret_cast<pthreadpool_t>(thread_pool_.get());
+}
```

Comment:
this thread pool is supposed to be created once and then accessed many times, we should think about an implementation where we don't have to take a lock each time we access the thread pool after it was created

Reply:
Considering the amount of work inside critical section (just read and cast a pointer) and the amount of work outside critical section (convolution), I'm not sure whether it's a premature optimization to use things like double checked locking and worry about memory barrier...